### PR TITLE
Made the RandomState conversion of Numba a valid RandomStateType

### LINKED ISSUE #661: Numba implementation of `RandomState` cannot update values in Graph.
## Description of your problem or feature request

**Please provide a minimal, self-contained, and reproducible example.**
```python
import aesara  
aesara.config.mode = "NUMBA"  
import aesara.tensor as at  
import numpy as np

rng = aesara.shared(np.random.RandomState(1))
rng_new = aesara.shared(np.random.RandomState(2))
x = at.random.normal(rng=rng)
# This will first throw an error in DeepCopyOp but that can be bypassed by removing the conditional check.
aesara.function([],x,updates={rng:rng_new})()  
```
The issue is that we essentially replace the `RandomStateType` of any given graph with Numba's internal state array in:

https://github.com/aesara-devs/aesara/blob/7c32fc48cc971889242c8a124ea577d8618f0ef2/aesara/link/numba/dispatch/random.py#L25-L30

However this state array isn't recognized by `RandomStateType.filter()` as a valid value. Hence it cannot update the rng with a new one.

### PR DESCRIPTION
**Pull request by kc611**
Resolves #661 

This PR makes the Numba conversion of `RandomStateType` a tuple representing Numba's internal state and is a valid `RandomStateType` even after conversion.

### CODE REVIEW

File: aesara/tensor/random/type.py [hunk 1]
```diff
@@ -70,6 +70,13 @@ def is_valid_value(a, strict):
             if state_key.shape == (624,) and state_key.dtype == np.uint32:
                 return True
 
+        # Workaround for supporting RandomState in Numba backend
+        if isinstance(a, tuple) and a[0] == "numba_internal_state_arr":
+            # The second element of tuple should be an array representing
+            # numba's internal state
+            if a[1].shape == (624,) and a[1].dtype == np.uint32:
+                return True
```
Comment (brandonwillard) [inline aesara/tensor/random/type.py:78]: Can we do this without making Numba-specific changes to generic types?

**Review by brandonwillard — CHANGES_REQUESTED** [review]

File: aesara/tensor/random/type.py (hunk 1 above)
Comment (kc611) [reply aesara/tensor/random/type.py]: The only way I could find to do this without making Numba-specific changes to generic types is that if we make `RandomState` an acceptable datatype in Numba. 

It doesn't necessarily need to hold all the information of the `RandomState` just be a 'fake' placeholder type just to get `RandomState` objects accepted in Numba and whenever it needs to be reconstructed as a `RandomState` (to get it out of the Numba function) we can do that using Numba's internal state array.

Comment (brandonwillard) [reply aesara/tensor/random/type.py]: Yeah, let's add a type for it to Numba; that sounds great.

**Review by brandonwillard — APPROVED** [review]
This looks great!  We're slowly getting closer to better and more direct Numba integration with things like this.

**Review by brandonwillard — CHANGES_REQUESTED** [review]
Before I forget, we should probably have a direct Numba unboxing test.  That will help simplify future work on this Numba typing code.  The current test is a little to indirect for easy development and debugging.

File: tests/link/test_numba.py [hunk 2]
```diff
@@ -293,6 +294,15 @@ def test_create_numba_signature(v, expected, force_scalar):
     assert res == expected
 
 
+@pytest.mark.parametrize(
+    "input, aesara_type, aesara_wrapper",
+    [(np.random.RandomState(1), RandomStateType, shared)],
+)
+def test_aesaratypes_box_unbox(input, aesara_type, aesara_wrapper):
+    res = function([], aesara_wrapper(input), mode=numba_mode)()
+    assert aesara_type.filter(res)
```
Comment (kc611) [inline tests/link/test_numba.py:303]: Doesn't look like the actual states stay consistent though. (looks like Numba's internal state changes, probably due to some unrelated operations, even if we aren't doing anything with it in the user side of the code.)

Normally doing something like this works:

```python
import numba
import numpy as np
from numba import _helperlib
rng = np.random.RandomState(1)
ints, index = rng.get_state()[1:3]
ptr = _helperlib.rnd_get_np_state_ptr()
_helperlib.rnd_set_state(ptr, (index, [int(x) for x in ints]))
state_list = _helperlib.rnd_get_state(_helperlib.rnd_get_np_state_ptr())[1]
assert np.all(ints == state_list) #Works
```

File: aesara/link/numba/dispatch/random.py [hunk 3]
```diff
@@ -22,12 +24,71 @@
...
+            # they can be accessed inside njit functions, if required.
+            ("state_key", types.Array(types.uint32, 1, "C")),
+        ]
+        models.StructModel.__init__(self, dmm, fe_type, members)
+
+
+@unbox(RandomStateNumbaType)
+def unbox_random_state(typ, obj, c):
+    """
+    Convert a RandomState object to a native RandomStateNumbaModel structure.
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch/random.py:55]:
```suggestion
    """Convert a `RandomState` object to a native `RandomStateNumbaModel` structure.
```

File: aesara/link/numba/dispatch/random.py [hunk 4]
```diff
@@ -22,12 +24,71 @@
...
+
+    Note that this will create a 'fake' structure which will just get the
+    RandomState objects accepted in Numba functions but the actual information
+    of the Numba's random state is stored internally and can be accessed
+    anytime by numba._helperlib.rnd_get_np_state_ptr()
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch/random.py:60]:
```suggestion
    `RandomState` objects accepted in Numba functions but the actual information
    of the Numba's random state is stored internally and can be accessed
    anytime using ``numba._helperlib.rnd_get_np_state_ptr()``.
```

File: aesara/link/numba/dispatch/random.py [hunk 5]
```diff
@@ -22,12 +24,71 @@
...
+    interval = cgutils.create_struct_proxy(typ)(c.context, c.builder)
+    is_error = cgutils.is_not_null(c.builder, c.pyapi.err_occurred())
+    return NativeValue(interval._getvalue(), is_error=is_error)
+
+
+@box(RandomStateNumbaType)
+def box_random_state(typ, val, c):
+    """
+    Convert a native RandomStateNumbaModel structure to an RandomState object
+    using Numba's internal state array.
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch/random.py:71]:
```suggestion
    """Convert a native `RandomStateNumbaModel` structure to an `RandomState` object
    using Numba's internal state array.
```

File: aesara/link/numba/dispatch/random.py [hunk 6]
```diff
@@ -22,12 +24,71 @@
...
+
+    Note that RandomStateNumbaModel is just a placeholder structure with no
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch/random.py:73]:
```suggestion
    Note that `RandomStateNumbaModel` is just a placeholder structure with no
```

File: aesara/link/numba/dispatch/random.py [hunk 7]
```diff
@@ -22,12 +24,71 @@
...
+    inherent information about Numba internal random state, all that information
+    is instead retrived from Numba using _helperlib.rnd_get_state() and a new
+    RandomState is constructed using the Numba's current internal state.
```
Comment (brandonwillard) [inline aesara/link/numba/dispatch/random.py:76]:
```suggestion
    is instead retrieved from Numba using ``_helperlib.rnd_get_state()`` and a new
    `RandomState` is constructed using the Numba's current internal state.
```

**Review by brandonwillard — CHANGES_REQUESTED** [review]
I meant we should test the boxing/unboxing on the Numba side _only_ (i.e. no Aesara involved).

File: tests/link/test_numba.py (hunk 2 above)
Comment (kc611) [reply tests/link/test_numba.py]: Nevermind, they're consistent now. 

I had the misconception that `RandomState(state_list)` is equivalent to `RandomState().set_state(("MT19937", state_list, pos))`, spoiler alert, it isn't. The former actually uses the array as seed to generate an entirely different state array.

**Review by brandonwillard — APPROVED** [review]

**Pull request merged**